### PR TITLE
Issue 5: improve new kill message

### PR DESCRIPTION
**Pull request by Veritania**
Targetting issue #5 

I have attempted to replace the raw text messages with Discord Embeds for each Killmail detected.

Note that I have upgraded the `discord.py (1.3.4 => 1.6.0)` and `aiohttp (3.5.4 => 3.6.0)` dependencies - I'm not sure if it's actually necessary.

### CODE REVIEW

**Comment by Veritania** [comment]
Currently I have an issue where it only generates 1 embed and posts it, even though it correctly fetches and iterates through up to <=25 of the latest killmails.

I'm not entirely sure why there's a hiccup, if it's a limit on discords side for embed posting or if there's some sort of race condition(?). It's the first time I'm working with python, so maybe someone can shed some light on it.

In it's current state, this is how the output embed looks: 
Discord_sjFabLvnnQ
**(Current look^)**

It does support listing multiple attackers, including if they have ECM/Drain/Supression. 
Here's an example from an older iteration, before I cleaned up the number output and removed the big thumbnail picture on the right side
Discord_YHoVWRgF2u
**(Outdated example^)**


I'd be happy to get some feedback on the layout, even though the embed is somewhat limited in how creative you can be when it comes to positioning the fields.

**Comment by Veritania** [comment]
The issue of only posting 1 embed has been solved with commit https://github.com/OpenPerpetuum/OPDiscordBot/pull/13/commits/d5467b6b530ad121486c48ee290f14e9dd8de737 🎉 
My lack of python skill was the issue, and I returned at the wrong point in the for loops :)

A few things remains for this to be released in my opinion, but in it's current state it will work - It just has some "beauty issues". 

I discovered that there is a limit on the amount of Fields that a single embed can have (25), and on Killmails with many attackers (or with attackers who use a lot of ECM/Drain etc) we quickly fill up that limit.

My suggestion is that in those scenarios we cut the list of attacks short, ideally sorted by who dealt the most damage and then specify that the full details are available in the Killboard Link. 
If we don't sort by damage, we might end up with only displaying a list of people who did 0 or very little damage, while the Victims 'damage taken' displays a much higher number. The Killing Blow should also have priority, for bragging rights :)

I would also prefer that when the bot is behind on multiple killmails, they get posted in correct order (sorted by ID). But I think that might be tricky since we're posting these messages with `await`, so it seems that the order gets messed up slightly.

I tested with iterating over 25 kills, and it seems to run 4-5 threads at a time trying to post a message, so in batches of 5 the order gets randomized.

**Comment by Veritania** [comment]
Example of how the killmail channel would be looking with the current changes, you can see a few of the messages gets a little busy with a lot of attackers/ECM/Drain
https://gfycat.com/shabbyglassdaddylonglegs

**Comment by Veritania** [comment]
I overhauled the process of building the discord embed
With commit https://github.com/OpenPerpetuum/OPDiscordBot/pull/13/commits/5213cd77d73838856cf01b60f9146af4acf57a74 the bot will now take into consideration if it's a large killmail. (Note that this commit is using test data to force reading a kill that would exceed the limit - It also contains a lot of debug print()s)

Instead of posting a killmail that gets cut off when we hit the Discord Limit of 25 Fields, it will now add as many full-detailed attackers as it can, and save the last Field for the "Too many attacks to list, see the Killboard Link".
It will also always list the Killing Blow Attacker on top (for bragging rights).

**Moving forward:**
Currently the list of attackers is not sorted by damage. This means the embed message might be populated with some entries that dealt 0 damage, and it leaves out entries that has dealt damage where we would cross the 25 Fields Limit.

I just learned that shield damage shows up as 0 damage on the API, and that's why some players appear where they seemingly have done nothing to participate in the kill (even as friendly fire!).

So I don't know if we *need* the Embed to be sorted by damage numbers or not, since it's not entirely true that damage is the only meaningful contribution and should be prioritized.

**Comment by Veritania** [comment]
Here is an example of a Killmail that would exceed the 25 Field Limit. The final field will tell people to go to the Killboard website and look at the details there.
Discord_TtN7KESFD6

**Comment by MikeJeffers** [comment]
Yeah maybe the killboard link for details might be more manageable, but knowing users they won't click the link.
What about multiple embeds?  Is there an issue with posting multiple embeds where you could do:
Kill summary embed
Kill details embed/Attacker list 0-25
Kill details embed/Attacker list 25-50
...
Kill details embed/Attacker list (N-25)-N

Big battles get a lot of attention and people like seeing them.  I know the data can be shown elsewhere, but it might be better if this was more standalone.

**Comment by MikeJeffers** [comment]
Also this will be a blocking issue for a killboard link: https://github.com/OpenPerpetuum/perpetuum-killboard/issues/4
Not even: https://killboard.byhub.ru/ (the better 3rd party kb) has a directly linkable view for kill details

**Comment by Veritania** [comment]
Splitting it into multiple embeds should be possible, but the challenge is to ensure that embed part 1 and embed part 2 is posted in sequence, and that it won't be broken up by another thread/coroutine trying to post a different kill.

It's already an issue we're running into now, if the bot is far behind on posting kills, the kills do not necessarily get sent in the order they are processed.

The discord library we are using to post messages, requires the await keyword because it's a coroutine -> https://discordpy.readthedocs.io/en/latest/api.html#discord.abc.Messageable.send

I currently don't know how I would go about tackling this threading scenario.

**Comment by Veritania** [comment]
I had no idea https://github.com/OpenPerpetuum/perpetuum-killboard/issues/4 was a thing, I even saw the comments at the top of the bot code which states `the OP killboard page doesn't allow for direct linking.` but I assumed it was an old issue since I didn't experience it.



It works on `Firefox 86.0.1` ✔️ , but not on Chromium based browsers for me.
Tested on:
`Google Chrome 89.0.4389.90` 🛑 
`Microsoft Edge 89.0.774.54` 🛑 

Example link used: https://killboard.openperpetuum.com/kill/2741 

pycharm64_WzZfbkltlM

**Comment by MikeJeffers** [comment]
> 
> 
> I had no idea [OpenPerpetuum/perpetuum-killboard#4](https://github.com/OpenPerpetuum/perpetuum-killboard/issues/4) was a thing, I even saw the comments at the top of the bot code which states `the OP killboard page doesn't allow for direct linking.` but I assumed it was an old issue since I didn't experience it.
> 
> It works on `Firefox 86.0.1` ✔️ , but not on Chromium based browsers for me.
> Tested on:
> `Google Chrome 89.0.4389.90` 🛑
> `Microsoft Edge 89.0.774.54` 🛑
> 
> Example link used: https://killboard.openperpetuum.com/kill/2741
> 
> pycharm64_WzZfbkltlM

Its a single-page react app, so its saving state in-memory.  Probably your browser is caching something, but if you hit the route /kill/:id the webserver serving does not send you to a page, nor does it properly send you to the root page and let react act like a "router" and dynamically repopulate the state and render the page as if it were a separate static page.
but thats an issue for another repo!

**Comment by MikeJeffers** [comment]
> 
> 
> Splitting it into multiple embeds should be possible, but the challenge is to ensure that embed part 1 and embed part 2 is posted in sequence, and that it won't be broken up by another thread/coroutine trying to post a different kill.
> 
> It's already an issue we're running into now, if the bot is far behind on posting kills, the kills do not necessarily get sent in the order they are processed.
> 
> The discord library we are using to post messages, requires the await keyword because it's a coroutine -> https://discordpy.readthedocs.io/en/latest/api.html#discord.abc.Messageable.send
> 
> I currently don't know how I would go about tackling this threading scenario.

Ah that's a good point! async/concurrency is at play here...

**Comment by Veritania** [comment]
An alternative idea to splitting the embed post, in case of too many attackers:

Instead of using a Field for each stat, maybe we should only list the Kill Target and Attackers as Fields, and make everything else part of that Fields descriptive text, similar to what it's currently doing when listing the Corp.

The formatting and visual will be a bit different, and there's a hard-limit on 1024 characters pr Field Description, with a total cap on 6000 characters for everything in the embed.

I'm estimating that we could list ~10ish Attackers with their stats, pr Field if we are okay with grouping them all up under a single field like that. (Otherwise <24 attackers, 1 in each remaining field)
Depending on how it counts the icons - `:crossed_swords:` ⚔️ and how long certain corp/player names are.

Example of how the Victim part would look with this change:
Discord_JjIndPWKjA

**Comment by Veritania** [comment]
I've set this up in another branch, that can be merged into this one if you like that approach better
https://github.com/Veritania/OPDiscordBot/tree/issue-5-experimental
Here's the way it ends up looking
Discord_iZ5TotmbTy

**Comment by Veritania** [comment]
Seems like we're mostly limited by the Embeds 6000 Char limit now with this approach.
Behold the maxed out kill report in all its glory
https://i.imgur.com/KvfwUvV.png 

Each "Attackers" section is limited to 1024 characters.
If we try to add another attackers full details and it would exceed that limit, we create a new "Attackers" section instead and put them there.


The characters spent pr Attacker varies quite a bit - `Name, Corp name, Killing Blow, ECM/Supress/Drain` all heavily influence how many entries an "Attackers" section can hold.
So once we're about to hit the 6000 total char limit, it will post the "Full details at Killboard link" section instead, and drop the remaining Attackers.

But seeing the size of this embed (with discord on 50% zoom only), I don't think it's reasonable to manage even larger kills. This is already overwhelming.

Feedback?  
Should I merge this into #13 and then we drop the "Field for every player stat" and instead build these Fields that contain bold text with newlines?

**Comment by MikeJeffers** [comment]
>  in another branch, that can be merged into this one if you like that appr

I like the compressed view much better.  Maximizing the limits of the embed to show more data where possible is best.  Showing big battles is even better.

**Comment by MikeJeffers** [comment]
> 
> 
> Seems like we're mostly limited by the Embeds 6000 Char limit now with this approach.
> Behold the maxed out kill report in all its glory
> https://i.imgur.com/KvfwUvV.png
> 
> Each "Attackers" section is limited to 1024 characters.
> If we try to add another attackers full details and it would exceed that limit, we create a new "Attackers" section instead and put them there.
> 
> The characters spent pr Attacker varies quite a bit - `Name, Corp name, Killing Blow, ECM/Supress/Drain` all heavily influence how many entries an "Attackers" section can hold.
> So once we're about to hit the 6000 total char limit, it will post the "Full details at Killboard link" section instead, and drop the remaining Attackers.
> 
> But seeing the size of this embed (with discord on 50% zoom only), I don't think it's reasonable to manage even larger kills. This is already overwhelming.
> 
> Feedback?
> Should I merge this into #13 and then we drop the "Field for every player stat" and instead build these Fields that contain bold text with newlines?

Yes
And remember, there is already a lot of enhancement the embed is providing here.  You are using whitespace/columnar templates to get more data on the screen than the simple text-based post was to begin with.
The icons and formatting is a bonus too.  It all looks 100x better!

And you are right, at the end of the day, we need to throw to the KB to show the full details in a scalable manner the user can better explore.  The fact you are even catching that case now is a big deal because I'm sure we didn't even check for max char length on our original code!

File: cogs/perpetuum_killboard_cog.py [hunk 1]
```diff
@@ -101,52 +158,88 @@ async def display_new_kills(self):
...
+            # Validate the rest of the attackers
+            for a in kill['_embedded']['attackers']:
+                current_atk_str = add_attacker_str(atk_saved_list, a)  # Add a new Attackers info to the Saved list
+                embed_length = len(kill_message_embed)  # Update current embed length
+                new_length = len(current_atk_str) + embed_length
+
+                if new_length >= 5800:  # Do we exceed the allowed Embed Length?
+                    # End early, build what we have and add 'too many attackers' field at the end of the embed
+                    build_attacker_field(atk_saved_list, kill_message_embed, atk_field_name)
+                    add_too_many_attackers(kill_message_embed)  # Costs 1 field + ~85 characters
```
Comment (Veritania) [inline cogs/perpetuum_killboard_cog.py:219]: _L219 & L233:_
This is a coroutine call, python would prefer if we used `await` when calling this.
However, the library we are using to interact with Discord does not allow for Embeds to be awaited - If you try you get:
`TypeError: object Embed can't be used in 'await' expression`

I have not run into any issues by *not* using `await` here, but your code editor might scream at you :)

File: cogs/perpetuum_killboard_cog.py [hunk 2]
```diff
@@ -101,52 +158,88 @@ async def display_new_kills(self):
...
+            atkcount = (str(len(kill['_embedded']['attackers'])))
+            if int(atkcount) <= 1:  # If there's only 1 attacker, change field name to singular.
+                atk_field_name = "⚔ Attacker"
+
+            atk_saved_list = ""  # Validated Attackers, saved to a list - These WILL exist in the embed
+            current_atk_str = ""  # Validated Attackers + New Attacker we want to add - Check if this exceeds limits
+            current_field_length = 0  # char-length of field we're currently manipulating
+            embed_length = len(kill_message_embed)  # Embed char length, MAX 6000
+            fields_built = len(kill_message_embed.fields)  # Keep track of total fields, MAX 25
+            inline_next_attacker_field = False
```
Comment (Veritania) [inline cogs/perpetuum_killboard_cog.py:198]: I thought it was more readable to understand the voodoo later, if these were initialized early and set to empty values with comments.
Technically not all are needed.

**Comment by Veritania** [comment]
This is a bit of an overhaul of the existing bots capability to post a kill message, some of the highlights are:

- We use Discords Embeds to contain a killboard message now, instead of posting raw text messages
Colors and Icons for visual improvement 🎉  (or impairment, personal preference)
- The Killed Player will always be listed first, followed by the Player who managed to get the killing blow!  
_Bragging rights are important!_
- The bot will list out the remaining Attackers
This is including if they managed to do any ECM/Surpressor/Drain numbers. A single kill message can support up to around 70 attackers, after that it will stop listing attackers and suggest people to look up the Killboard instead
- Each Kill message features a hyperlink at the top, that will bring you to the specific kill on https://killboard.openperpetuum.com/ (Need to solve OpenPerpetuum/perpetuum-killboard#4 for that to work 100%)

The final iteration here has removed the cartoony skull icons used in previous examples.
This is what a typical killmail will look like: https://i.imgur.com/MpUOvQo.png

**Review by MikeJeffers — APPROVED** [review]
LGTM

**Pull request merged**